### PR TITLE
Lower the log level of the expected exceptions to DEBUG

### PR DESCRIPTION
**Pull request by minwoox**
Motivation:
Even though the expected exceptions during unit tests are literally expected, they are shown as
`WARN` level so can give some confuse to us.

Modification:
- Add `(expected exception)` and  lower the log level to `DEBUG` using `ExpectedExceptionAppender`

Result:
- Can distinguish the expected exceptions from unexpected ones

### CODE REVIEW

File: testing-internal/src/main/java/com/linecorp/centraldogma/testing/internal/ExpectedExceptionAppender.java [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+        }
+
+        @Override
+        public Object[] getArgumentArray() {
+            return event.getArgumentArray();
+        }
+
+        @Override
+        public Level getLevel() {
+            return Level.DEBUG;
```
Comment (trustin) [inline testing-internal/src/main/java/com/linecorp/centraldogma/testing/internal/ExpectedExceptionAppender.java]: How about returning `event.getLevel()` if it's lower than `Level.DEBUG`?

File: testing-internal/src/main/java/com/linecorp/centraldogma/testing/internal/ExpectedExceptionAppender.java [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+     * {@code try-with-resources} block:
+     * <pre>{@code
+     * try (SafeCloseable ignored = ExpectedExceptionAppender.appendIfMatch(throwable, cause)) {
+     *     ...
+     *     // Expected exception happens.
+     *     ...
+     * }
+     * }</pre>
+     */
+    public static SafeCloseable appendIfMatch(Class<?> throwable, String cause) {
```
Comment (trustin) [inline testing-internal/src/main/java/com/linecorp/centraldogma/testing/internal/ExpectedExceptionAppender.java]: `appendIfMatch` makes me feel like the event will not be appended if not matches. How about: `withExpectedException()`? This also makes me think the following pattern may be better because `SafeCloseable` is unnecessary completely.

```java
withExpectedException(SomeException.class, "text", () -> {
    assertThatThrownBy(...);
});
```

Comment (minwoox) [reply testing-internal/src/main/java/com/linecorp/centraldogma/testing/internal/ExpectedExceptionAppender.java]: That's a good idea~!! Thank you~!

File: testing-internal/src/main/java/com/linecorp/centraldogma/testing/internal/ExpectedExceptionAppender.java (hunk 1 above)
Comment (minwoox) [reply testing-internal/src/main/java/com/linecorp/centraldogma/testing/internal/ExpectedExceptionAppender.java]: Changed it~!!

File: testing-internal/src/main/java/com/linecorp/centraldogma/testing/internal/ExpectedExceptionAppender.java [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+        } finally {
+            exceptionAndMessage.remove(throwable);
+        }
+    }
+
+    @Override
+    protected void append(ILoggingEvent eventObject) {
+        final IThrowableProxy exception = getException(eventObject);
+            for (String exceptionName : exceptionAndMessage.keySet()) {
+            if (exception != null && exception.getClassName().endsWith(exceptionName) &&
```
Comment (hyangtack) [inline testing-internal/src/main/java/com/linecorp/centraldogma/testing/internal/ExpectedExceptionAppender.java]: indent?

File: testing-internal/src/main/java/com/linecorp/centraldogma/testing/internal/ExpectedExceptionAppender.java [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+     * Adds {@code "(expected exception)"} to the log message and lower its logging level to
+     * {@link Level#DEBUG} if the exception occurred from the {@code action} matches {@code throwable},
+     * and contains {@code message}.
+     */
+    public static void withExpectedException(String throwable, String message,
+                                             ThrowingRunnable action) throws Exception {
+        requireNonNull(throwable, "throwable");
+        requireNonNull(message, "message");
+        requireNonNull(action, "action");
+        exceptionAndMessage.put(throwable, message);
```
Comment (hyangtack) [inline testing-internal/src/main/java/com/linecorp/centraldogma/testing/internal/ExpectedExceptionAppender.java]: Just curious. Is there any chance to put messages with the same `throwable` at the same time?

Comment (minwoox) [reply testing-internal/src/main/java/com/linecorp/centraldogma/testing/internal/ExpectedExceptionAppender.java]: I think it's OK~! :-)
I was intending to use this only in the scope of `run()` in `ThrowingRunnable`. If the code emits more than one expected exception, then it must be separated in different `run()`. This is used only in unit tests and it is executed sequentially, so it's ok.

File: testing-internal/src/main/java/com/linecorp/centraldogma/testing/internal/ExpectedExceptionAppender.java (hunk 3 above)
Comment (minwoox) [reply testing-internal/src/main/java/com/linecorp/centraldogma/testing/internal/ExpectedExceptionAppender.java]: oops~!!

File: it/src/test/java/com/linecorp/centraldogma/it/GetFileTest.java [hunk 5]
```diff
@@ -42,34 +46,47 @@ public void testInvalidJsonPath() throws TException {
         assertThatThrownBy(() -> rule.client().getFile(
                 rule.project(), rule.repo1(), Revision.HEAD,
                 Query.ofJsonPath("/test/test2.json", "$.non_exist_path")).join())
-                .isInstanceOf(CompletionException.class).hasCauseInstanceOf(CentralDogmaException.class)
+                .isInstanceOf(CompletionException.class).hasCauseInstanceOf(
+                CentralDogmaException.class)
```
Comment (trustin) [inline it/src/test/java/com/linecorp/centraldogma/it/GetFileTest.java]: Seems unrelated?

File: it/src/test/java/com/linecorp/centraldogma/it/GetFileTest.java [hunk 6]
```diff
@@ -42,34 +46,47 @@ public void testInvalidJsonPath() throws TException {
...
-    public void testInvalidFile() throws TException {
-        assertThatThrownBy(() -> rule.client().getFile(
-                rule.project(), rule.repo1(), Revision.HEAD,
-                Query.ofJsonPath("/test/non_existing_file.json", "$.a")).join())
-                .isInstanceOf(CompletionException.class).hasCauseInstanceOf(CentralDogmaException.class)
-                .matches(e -> ((CentralDogmaException) e.getCause()).getErrorCode() == ENTRY_NOT_FOUND);
+    public void testInvalidFile() throws Exception {
+        ExpectedExceptionAppender.withExpectedException(
+                EntryNotFoundException.class, "non_existing_file", () ->
+                        assertThatThrownBy(() -> rule.client().getFile(
```
Comment (trustin) [inline it/src/test/java/com/linecorp/centraldogma/it/GetFileTest.java]: It seems like it's fairly common to call a single `assertThatThrownBy()` with `withExpectedException`. How about adding this:

```java
import static ...ExpectedExceptionAppender.assertThatThrownByWithExpectedException;
assertThatThrownByWithExpectedException(
        EntryNotFoundException.class, "non_existing_file",
        () -> rule.client().....).isInstanceOf(...);
```

File: it/src/test/java/com/linecorp/centraldogma/it/WatchTest.java [hunk 7]
```diff
@@ -79,14 +78,19 @@ public void testWatchRepository() throws Exception {
         final Revision rev2 = commit.revision();
 
         assertThat(rev2).isEqualTo(rev1.forward(1));
-        assertThat(future.get(3, TimeUnit.SECONDS)).isEqualTo(rev2);
+        ExpectedExceptionAppender.withExpectedException(
+                "AnnotatedConnectException", "Connection refused: /127.0.0.1:", () ->
```
Comment (trustin) [inline it/src/test/java/com/linecorp/centraldogma/it/WatchTest.java]: 'Connection refused' and IP address are platform-dependent strings. Could we only check the exception type?

File: testing-internal/src/main/java/com/linecorp/centraldogma/testing/internal/ExpectedExceptionAppender.java (hunk 4 above)
Comment (hyangtack) [reply testing-internal/src/main/java/com/linecorp/centraldogma/testing/internal/ExpectedExceptionAppender.java]: Ok! Thank you for explanation! :-)

File: it/src/test/java/com/linecorp/centraldogma/it/GetFileTest.java (hunk 6 above)
Comment (minwoox) [reply it/src/test/java/com/linecorp/centraldogma/it/GetFileTest.java]: That's a brilliant idea~! Thank you~!!

**Comment by trustin** [comment]
Thanks, @minwoox !
